middle_end/flambda/types/env/aliases.ml

```diff
@@ -446,174 +647,202 @@ let mem t element =
-            simples
-      in
-      match
-        Aliases_of_canonical_element.find_earliest_candidates aliases
-          ~filter_by_scope ~min_name_mode
-      with
-      | Some at_earliest_mode ->
-        (* Aliases_of_canonical_element.find_earliest_candidates only returns
-           non-empty sets *)
-        assert (not (Simple.Set.is_empty at_earliest_mode));
-        Simple.Set.fold (fun elt min_elt ->
-            if defined_earlier t elt ~than:min_elt
-            then elt
-            else min_elt)
-          at_earliest_mode
-          (Simple.Set.min_elt at_earliest_mode)
-      | None -> raise Not_found
+  let find_earliest () =
+    (* There used to be a shortcut that avoided consulting the aliases in the
+       common case that [element] is itself canonical and has no aliases, since
+       then it does not appear in [canonical_elements]. However, this shortcut
+       was broken: a canonical element *with* known aliases may still not appear
+       in [canonical_elements]. See tests/flambda2-aliases for a test that gave
+       incorrect output (saying x/39 had no aliases). It may be worth restoring
+       the shortcut, perhaps by returning more information from [canonical]. *)
+    let aliases = get_aliases_of_canonical_element t ~canonical_element in
+    let filter_by_scope name_mode names =
+      if Name_mode.equal name_mode Name_mode.in_types then names
+      else
+        Name.Set.filter (fun name ->
+            let binding_time_and_mode =
+              Name.Map.find name t.binding_times_and_modes
+            in
+            let scoped_name_mode =
+              Binding_time.With_name_mode.scoped_name_mode
+                binding_time_and_mode
+                ~min_binding_time
+            in
+            Name_mode.equal name_mode scoped_name_mode)
+          names
     in
     match
-      Name_mode.compare_partial_order
-        (name_mode t canonical_element ~min_binding_time)
-        min_name_mode
+      Aliases_of_canonical_element.find_earliest_candidates aliases
+        ~filter_by_scope ~min_name_mode
     with
-    | None -> find_earliest ()
-    | Some c ->
-      if c >= 0 then canonical_element
-      else find_earliest ()
+    | Some at_earliest_mode ->
+      (* Aliases_of_canonical_element.find_earliest_candidates only returns
+         non-empty sets *)
+      assert (not (Name.Set.is_empty at_earliest_mode));
+      Name.Set.fold (fun elt min_elt ->
+          if name_defined_earlier t elt ~than:min_elt
+          then elt
+          else min_elt)
+        at_earliest_mode
+        (Name.Set.min_elt at_earliest_mode)
+      |> Simple.name
+    | None -> raise Not_found
+  in
+  match
+    Name_mode.compare_partial_order name_mode min_name_mode
+  with
+  | None -> find_earliest ()
+  | Some c ->
+    if c >= 0 then canonical_element
+    else find_earliest ()
 
 let get_aliases t element =
+  let create_alias_set ~canonical_element ~names =
```

Comment:
Could you move [create_alias_set] to [Alias_set] directly?

Reply:
Good idea, done